qiskit/circuit/singleton_gate.py

```diff
@@ -191,3 +221,220 @@ def copy(self, name=None):
         if not self.mutable and name is None:
             return self
         return super().copy(name=name)
+
+
+class SingletonControlledGate(ControlledGate):
+    """A base class to use for ControlledGate objects that by default are singleton instances
+
+    This class should be used for gate classes that have fixed definitions and
+    do not contain any unique state by default. The excpetion being if a custom control
+    state is set. The canonical example of something like this is :class:`~.CHGate` which
+    has an immutable definition and any instance of :class:`~.CHGate` without a ``ctrl_state``
+    set is the same. Using singleton gate enables using as a base class for these types of gate
+    classes it provides a large advantage in the memory footprint.
+
+    The exception to be aware of with this class though are the :class:`~.ControlledGate`
+    attributes :attr:`.label`, :attr:`.condition`, :attr:`.duration`, :attr`.unit`,
+    :attr:`ctrl_state` which can be set differently for specific instances of gates.
+    For :class:`~.SingletonControlledGate` usage to be sound setting these attributes
+    is not available and they can only be set at creation time. If any of these
+    attributes are used instead of using a single shared global instance of
+    the same gate a new separate instance will be created.
+    """
+
+    _instance = None
+
+    def __new__(cls, *args, **kwargs):
+        if args or (  # pylint: disable=too-many-boolean-expressions
+            kwargs
+            and (
+                kwargs.get("label", None) is not None
```

Comment:
After the last rebase this is using the machinery introduced in https://github.com/Qiskit/qiskit/pull/11014 none of this logic to control creating singleton instances exists anymore.

FWIW, the reason I did it this way was to try and optimize for speed so that the if statement would bail out as early as possible. The overhead of the old structure was a lot higher so any time spent evaluating the if statements here made it really slow in aggregate to create new gate objects.